contracts/SP1_SpokePool.sol

```diff
@@ -0,0 +1,168 @@
+    // the input data was published on L1 by the HubPool. This input data is then executed on this contract.
+    // This modifier sets the adminCallValidated variable so this condition can be checked in _requireAdminSender().
+    modifier validateInternalCalls() {
+        // Make sure adminCallValidated is set to True only once at beginning of the function, which prevents
+        // the function from being re-entered.
+        if (_adminCallValidated) {
+            revert AdminCallAlreadySet();
+        }
+
+        // This sets a variable indicating that we're now inside a validated call.
+        // Note: this is used by other methods to ensure that this call has been validated by this method and is not
+        // spoofed.
+        _adminCallValidated = true;
+
+        _;
+
+        // Reset adminCallValidated to false to disallow admin calls after this method exits.
+        _adminCallValidated = false;
+    }
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor(
+        address _verifier,
+        address _helios,
+        bytes32 _acrossCallProgramVKey,
+        address _hubPool,
+        address _wrappedNativeTokenAddress,
+        uint32 _depositQuoteTimeBuffer,
+        uint32 _fillDeadlineBuffer
+    ) SpokePool(_wrappedNativeTokenAddress, _depositQuoteTimeBuffer, _fillDeadlineBuffer) {
+        verifier = _verifier;
+        helios = _helios;
+        acrossCallProgramVKey = _acrossCallProgramVKey;
+        hubPool = _hubPool;
+    }
+
+    /**
+     * @notice This can be called by an EOA to relay call data stored on the HubPool on L1 into this contract.
+     * @dev Consider making this an onlyOwner function so that only a priveleged EOA can relay state. This EOA
+     * wouldn't be able to tamper with the _publicValues.
+     * @param _publicValues L1 contract state we want to relay into this contract. Contains a message that will
+     * be treated as calldata for a delegatecall into this contract.
+     * @param _proofBytes Proof bytes for the public values.
+     * @param _head Head number related to _publicValues.
+     */
+    function receiveL1State(
+        bytes calldata _publicValues,
+        bytes calldata _proofBytes,
+        uint256 _head
+    ) external validateInternalCalls {
+        // Verify proof and public values match:
+        ISP1Verifier(verifier).verifyProof(acrossCallProgramVKey, _publicValues, _proofBytes);
+        ContractPublicValues memory publicValues = abi.decode(_publicValues, (ContractPublicValues));
+
+        // Verify Helios light client is aware of the state root containing the public values:
+        bytes32 executionStateRoot = IHelios(helios).executionStateRoots(_head);
+        if (executionStateRoot != publicValues.stateRoot) {
+            revert StateRootMismatch();
+        }
+
+        // Validate state is intended to be sent to this contract:
+        (address _hubPool, address _target, bytes memory _message) = abi.decode(
```

Comment:
Would it be the `DataStore` address that we should be validating against since that's where the data is stored on mainnet (not the `HubPool`)?

Reply:
Yeah same page. See https://github.com/across-protocol/contracts/pull/916#discussion_r1990501728